types/json_path_expr.go

```diff
@@ -50,34 +50,89 @@ import (
 		select json_extract('{"a": "b", "c": [1, "2"]}', '$.*') -> ["b", [1, "2"]]
 */
 
+// if index is positive, it represents the [index]
+// if index is negative, it represents the [len() + index]
+// a normal index "5" will be parsed into "5", and the "last - 5" will be "-6"
+type jsonPathArrayIndex int
+
+func (index jsonPathArrayIndex) getIndexFromStart(obj BinaryJSON) int {
+	if index < 0 {
+		return obj.GetElemCount() + int(index)
+	}
+
+	return int(index)
+}
+
+// validateIndexRange returns whether `a` could be less or equal than `b`
+// if two indexes are all non-negative, or all negative, the comparison follows the number order
+// if the sign of them differs, this function will still return true
+func validateIndexRange(a jsonPathArrayIndex, b jsonPathArrayIndex) bool {
+	if (a >= 0 && b >= 0) || (a < 0 && b < 0) {
+		return a <= b
+	}
+
+	return true
+}
+
+func jsonPathArrayIndexFromStart(index int) jsonPathArrayIndex {
+	return jsonPathArrayIndex(index)
+}
+
+func jsonPathArrayIndexFromLast(index int) jsonPathArrayIndex {
+	return jsonPathArrayIndex(-1 - index)
+}
+
+type jsonPathArraySelection interface {
+	jsonPathArraySelection()
+}
+
+var _ jsonPathArraySelection = jsonPathArraySelectionAsterisk{}
+var _ jsonPathArraySelection = jsonPathArraySelectionIndex{}
+var _ jsonPathArraySelection = jsonPathArraySelectionRange{}
+
+type jsonPathArraySelectionAsterisk struct{}
+
+func (jsonPathArraySelectionAsterisk) jsonPathArraySelection() {}
+
+type jsonPathArraySelectionIndex struct {
+	index jsonPathArrayIndex
+}
+
+func (jsonPathArraySelectionIndex) jsonPathArraySelection() {}
+
+// jsonPathArraySelectionRange represents a closed interval
+type jsonPathArraySelectionRange struct {
+	start jsonPathArrayIndex
+	end   jsonPathArrayIndex
+}
+
+func (jsonPathArraySelectionRange) jsonPathArraySelection() {}
+
 type jsonPathLegType byte
 
 const (
 	// jsonPathLegKey indicates the path leg with '.key'.
 	jsonPathLegKey jsonPathLegType = 0x01
-	// jsonPathLegIndex indicates the path leg with form '[number]'.
-	jsonPathLegIndex jsonPathLegType = 0x02
+	// jsonPathLegArraySelection indicates the path leg with form '[index]', '[index to index]'.
+	jsonPathLegArraySelection jsonPathLegType = 0x02
 	// jsonPathLegDoubleAsterisk indicates the path leg with form '**'.
 	jsonPathLegDoubleAsterisk jsonPathLegType = 0x03
 )
 
 // jsonPathLeg is only used by JSONPathExpression.
 type jsonPathLeg struct {
-	typ        jsonPathLegType
-	arrayIndex int    // if typ is jsonPathLegIndex, the value should be parsed into here.
-	dotKey     string // if typ is jsonPathLegKey, the key should be parsed into here.
+	typ            jsonPathLegType
+	arraySelection jsonPathArraySelection // if typ is jsonPathLegArraySelection, the value should be parsed into here.
```

Comment:
please be aware that making `jsonPathLeg` can be copied correctly, we have a function `JSONPathExpression.clone()`

Reply:
Thanks for the reminder! This interface is implemented with the value as receiver, so it's always copied :thinking: .